staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -418,3 +439,246 @@ func TestRouting(t *testing.T) {
 		})
 	}
 }
+
+func TestHandlerConversionWithWatchCache(t *testing.T) {
+	testHandlerConversion(t, true)
+}
+
+func TestHandlerConversionWithoutWatchCache(t *testing.T) {
+	testHandlerConversion(t, false)
+}
+
+func testHandlerConversion(t *testing.T, enableWatchCache bool) {
+	cl := fake.NewSimpleClientset()
+	informers := informers.NewSharedInformerFactory(fake.NewSimpleClientset(), 0)
+	crdInformer := informers.Apiextensions().V1().CustomResourceDefinitions()
+
+	server, storageConfig := etcd3testing.NewUnsecuredEtcd3TestClientServer(t)
+	defer server.Terminate(t)
+
+	crd := multiVersionFixture.DeepCopy()
+	if _, err := cl.ApiextensionsV1().CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{}); err != nil {
+		t.Fatal(err)
+	}
+	if err := crdInformer.Informer().GetStore().Add(crd); err != nil {
+		t.Fatal(err)
+	}
+
+	etcdOptions := options.NewEtcdOptions(storageConfig)
+	etcdOptions.StorageConfig.Codec = unstructured.UnstructuredJSONScheme
+	restOptionsGetter := generic.RESTOptions{
+		StorageConfig:           &etcdOptions.StorageConfig,
+		Decorator:               generic.UndecoratedStorage,
+		EnableGarbageCollection: true,
+		DeleteCollectionWorkers: 1,
+		ResourcePrefix:          crd.Spec.Group + "/" + crd.Spec.Names.Plural,
+		CountMetricPollPeriod:   time.Minute,
+	}
+	if enableWatchCache {
+		restOptionsGetter.Decorator = genericregistry.StorageWithCacher(100)
+	}
+
+	handler, err := NewCustomResourceDefinitionHandler(
+		&versionDiscoveryHandler{}, &groupDiscoveryHandler{},
+		crdInformer,
+		http.HandlerFunc(func(http.ResponseWriter, *http.Request) {}),
+		restOptionsGetter,
+		dummyAdmissionImpl{},
+		&establish.EstablishingController{},
+		dummyServiceResolverImpl{},
+		func(r webhook.AuthenticationInfoResolver) webhook.AuthenticationInfoResolver { return r },
+		1,
+		dummyAuthorizerImpl{},
+		time.Minute, time.Minute, nil, 3*1024*1024)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	crdInfo, err := handler.getOrCreateServingInfoFor(crd.UID, crd.Name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	updateValidateFunc := func(ctx context.Context, obj, old runtime.Object) error { return nil }
+	validateFunc := func(ctx context.Context, obj runtime.Object) error { return nil }
+	startResourceVersion := ""
+
+	if enableWatchCache {
+		// Let watch cache establish initial list
+		time.Sleep(time.Second)
```

Comment:
Is this long enough to not flake in CI (same comment for all the sleeps)?

Reply:
we're the only reader/writer in this test, and I didn't want to slow unit tests down more than that, so this feels generous. if the watch cache is really slow for some reason, it will fall through to underlying storage and the test will still pass, not flake.